synchronizer/synchronizer.go

```diff
@@ -339,31 +371,54 @@ func (s *Synchronizer) AdvanceView(syncInfo hotstuff.SyncInfo) {
 
 	leader := s.leaderRotation.GetLeader(s.currentView)
 	if leader == s.opts.ID() {
-		s.consensus.Propose(syncInfo)
+		s.consensus.Propose(syncInfo.WithQC(s.highQC))
 	} else if replica, ok := s.configuration.Replica(leader); ok {
 		replica.NewView(syncInfo)
 	}
 }
 
+// isInPipelineStretch checks wether the given view lies
+// less then the pipelinestretch from the highQC
+func (s *Synchronizer) isInPipelineStretch(v hotstuff.View) bool {
+	return v > s.highQC.View() && v < s.highQC.View()+s.pipelinedViews
+}
+
 // updateHighQC attempts to update the highQC, but does not verify the qc first.
 // This method is meant to be used instead of the exported UpdateHighQC internally
 // in this package when the qc has already been verified.
+// This method ensures, the block of the highQC is always available locally.
 func (s *Synchronizer) updateHighQC(qc hotstuff.QuorumCert) {
-	newBlock, ok := s.blockChain.Get(qc.BlockHash())
-	if !ok {
-		s.logger.Info("updateHighQC: Could not find block referenced by new QC!")
-		return
+	if qc.View() > s.highQC.View() {
+		highQCBlock, ok := s.blockChain.Get(s.highQC.BlockHash())
+		if !ok {
+			s.logger.Info("updateHighQC: Could not find block referenced by new QC!")
+			return
+		}
+
+		s.highQC = qc
+		s.logger.Debug("HighQC updated")
+
+		if s.leafBlock.View() < s.highQC.View() || !s.blockChain.Extends(s.leafBlock, highQCBlock) {
+			s.leafBlock = highQCBlock
+		}
 	}
+}
 
-	oldBlock, ok := s.blockChain.Get(s.highQC.BlockHash())
+// updateLeafBlock attempts to update the leafBlock.
+// This method ensures, leafblock extends highQC.
```

Comment:
leaf block

Reply:
Also remove comma

Reply:
Fixed in [df3fc32](https://github.com/relab/hotstuff/pull/78/commits/df3fc324538afeeadc39d8adedcddb7d618f4ee8)